### PR TITLE
[WIP] add proto definition for cloud drive SDK

### PR DESCRIPTION
**Pull request by harsh-px**
Signed-off-by: Harsh Desai <harsh@portworx.com>

**What this PR does / why we need it**: 

This adds the SDK interface for cloud drive operations. 

**Special notes for your reviewer**:

### CODE REVIEW

File: api/api.proto [hunk 1]
```diff
@@ -232,6 +232,32 @@ message StoragePool {
...
+	// Category indicates how this drive is being used by the storage plugin
+	DriveCategory category = 6;
+}
+
+// DriveCategory is the type for various categories a drive could belong to
+enum DriveCategory {
+  // Data indicates drive is being used for data purposes
+  data = 0;
+  // Metadata indicates drive is being used for metadata purposes
+  metadata = 1;
```
Comment (adityadani) [inline api/api.proto]: journal is missing

File: api/api.proto [hunk 2]
```diff
@@ -2786,10 +2814,10 @@ message SdkCloudBackupCreateRequest {
   // Labels are list of key value pairs to tag the cloud backup. These labels
   // are stored in the metadata associated with the backup.
   map<string, string> labels = 5;
-  // FullBackupFrequency indicates number of incremental backup after whcih 
+  // FullBackupFrequency indicates number of incremental backup after whcih
```
Comment (adityadani) [inline api/api.proto]: nit: which

File: api/clouddrive.proto [hunk 3]
```diff
@@ -0,0 +1,232 @@
...
+
+  // Delete inspects the given drive set
+  rpc Inspect(SdkCloudDriveSetInspectRequest)
+    returns (SdkCloudDriveSetInspectResponse) {
+      option(google.api.http) = {
+        get: "/v1/clouddrives/{id}"
+      };
+    }
+
+  // Attach attaches all the drives in the given drive set
```
Comment (adityadani) [inline api/clouddrive.proto]: nit: Attach attaches all the drives in the given drive set to a single instance provided in the request

File: api/clouddrive.proto [hunk 4]
```diff
@@ -0,0 +1,232 @@
...
+  // Type is the type of the CloudDrive
+  string type = 2;
+  // Options are other opaque key value pair options
+  map<string, string> options = 3;
+}
+
+// Defines a response with a list of cloud drives
+message SdkCloudDriveSetCreateResponse {
+  // List of all the cloud drive ids in the cluster
+  repeated CloudDriveSet cloud_drives = 1;
```
Comment (adityadani) [inline api/clouddrive.proto]: this should be a single instance of CloudDriveSet instead of repeated

File: api/clouddrive.proto [hunk 5]
```diff
@@ -0,0 +1,232 @@
...
+}
+
+// Empty request
+message SdkCloudDriveSetEnumerateRequest {
+}
+
+// Defines a response with a list of cloud drives
+message SdkCloudDriveSetEnumerateResponse {
+  // List of all the cloud drive ids in the cluster
+  repeated CloudDriveSet cloud_drives = 1;
```
Comment (adityadani) [inline api/clouddrive.proto]: shouldn't this be `repeated strings` ?

File: api/clouddrive.proto [hunk 6]
```diff
@@ -0,0 +1,232 @@
...
+
+}
+
+// Empty response
+message SdkCloudDriveSetDetachResponse {
+}
+
+// Defines a request object to expand a cloud drive
+message SdkCloudDriveSetExpandRequest {
+  // ID of the cloud drive to expand
```
Comment (adityadani) [inline api/clouddrive.proto]: nit: ID of the cloud drive **set** to expand

File: api/api.proto (hunk 2 above)
Comment (jvinod) [inline api/api.proto]: ~whcih~ which

File: api/clouddrive.proto [hunk 7]
```diff
@@ -0,0 +1,232 @@
...
+
+option go_package = "api";
+option java_multiple_files = true;
+option java_package = "com.openstorage.api";
+
+// CloudDriveSet defines a set of drive pools that could be attached on a node
+message CloudDriveSet {
+	// ID is the unique id of the cloud drive. This is managed by the storage provider
+  // implementing the handlers
+	string id = 1;
```
Comment (adityadani) [inline api/clouddrive.proto]: May be we can be explicit here and say that this id is the nodeID. This makes it clear that a single CloudDriveSet can be associated with a node.

File: api/api.proto [hunk 8]
```diff
@@ -232,6 +232,32 @@ message StoragePool {
...
+	string id = 3;
+	// Path is the path where the drive is attached
+	string path = 4;
+	// Iops is the iops that the drive supports
+	int64 iops = 5;
+	// Category indicates how this drive is being used by the storage plugin
+	DriveCategory category = 6;
+}
+
+// DriveCategory is the type for various categories a drive could belong to
```
Comment (jvinod) [inline api/api.proto]: How do we use this?

File: api/clouddrive.proto [hunk 9]
```diff
@@ -0,0 +1,232 @@
...
+import "google/api/annotations.proto";
+import "api/api.proto";
+
+package openstorage.api;
+
+option go_package = "api";
+option java_multiple_files = true;
+option java_package = "com.openstorage.api";
+
+// CloudDriveSet defines a set of drive pools that could be attached on a node
```
Comment (jvinod) [inline api/clouddrive.proto]: set of StoragePool

File: api/clouddrive.proto [hunk 10]
```diff
@@ -0,0 +1,232 @@
...
+  // implementing the handlers
+	string id = 1;
+	// Pools describes the configuration of the drives present in this set
+	repeated StoragePool pools = 2;
+	// CreateTimestamp is the timestamp when the drive set was created
+  google.protobuf.Timestamp create_timestamp = 3;
+	// State state of the drive set from the well defined states
+	CloudDriveSetState state = 4;
+	// SchedulerNodeName is the name of the node in scheduler context who is using
+  // the cloud drive set
```
Comment (jvinod) [inline api/clouddrive.proto]: instance_name or eliminate this field -

File: api/clouddrive.proto [hunk 11]
```diff
@@ -0,0 +1,232 @@
...
+	string instance_id = 6;
+	// Zone defines the zone in which the node exists
+	string zone = 7;
+}
+
+
+// CloudDriveSetState indicates the state of the DriveSet
+enum CloudDriveSetState {
+	// Initializing indicates that the DriveSet is attached to a node but still in
+  // init state
```
Comment (jvinod) [inline api/clouddrive.proto]: What is this state

File: api/clouddrive.proto [hunk 12]
```diff
@@ -0,0 +1,232 @@
...
+	initializing = 0;
+	// InUse indicates that the DriveSet is in use
```
Comment (jvinod) [inline api/clouddrive.proto]: 'attached' or 'detached' ?

File: api/clouddrive.proto [hunk 13]
```diff
@@ -0,0 +1,232 @@
...
+	in_use = 1;
+	// PendingDecommission indicates that the DriveSet needs to be deleted.
+	// TODO: Currently this is being used only for storage less drive sets
```
Comment (jvinod) [inline api/clouddrive.proto]: 'deleted'

File: api/clouddrive.proto (hunk 3 above)
Comment (jvinod) [inline api/clouddrive.proto]: Attach() should take in an instance ID

File: api/clouddrive.proto [hunk 14]
```diff
@@ -0,0 +1,232 @@
...
+    returns (SdkCloudDriveSetAttachResponse) {
+      option(google.api.http) = {
+        post: "/v1/clouddrives/{id}/attach"
+        body: "*"
+      };
+    }
+
+  // Detach detaches all the drives in the given drive set from the instance
+  // where it's currently attached
+  rpc Detach(SdkCloudDriveSetDetachRequest)
```
Comment (jvinod) [inline api/clouddrive.proto]: Detach should also take in an instance ID

File: api/clouddrive.proto (hunk 3 above)
Comment (harsh-px) [reply api/clouddrive.proto]: it already does. It's in SdkCloudDriveSetAttachRequest.

File: api/clouddrive.proto (hunk 14 above)
Comment (harsh-px) [reply api/clouddrive.proto]: It already does. It's in SdkCloudDriveSetDetachResponse.

File: api/clouddrive.proto (hunk 5 above)
Comment (adityadani) [reply api/clouddrive.proto]: I think then the comment needs to change

File: api/clouddrive.proto [hunk 15]
```diff
@@ -0,0 +1,233 @@
...
+        get: "/v1/clouddrives/{id}"
+      };
+    }
+
+  // Attach attaches all the drives in the given drive set to a single instance
+  // provided in the request
+  rpc Attach(SdkCloudDriveSetAttachRequest)
+    returns (SdkCloudDriveSetAttachResponse) {
+      option(google.api.http) = {
+        post: "/v1/clouddrives/{id}/attach"
```
Comment (piyush-nimbalkar) [inline api/clouddrive.proto]: I feel this should attach/detach should be put operations by RESTful ways as they indicate modification of the drives, but no strong preference.

File: api/clouddrive.proto [hunk 16]
```diff
@@ -0,0 +1,233 @@
...
+	attached = 1;
+	// Detached indicates that the DriveSet is detached and not used by any node
+	detached = 2;
+	// PendingDecommission indicates that the DriveSet needs to be deleted
+	pending_decommission = 3;
+}
+
+// OpenStorageCloudDrives is a service used to manage cloud drives in the cluster
+service OpenStorageCloudDrives {
+  // Enumerate returns the ids of all the cloud drives in the current cluster
```
Comment (piyush-nimbalkar) [inline api/clouddrive.proto]: nit: *returns the list of cloud drive sets

File: api/clouddrive.proto [hunk 17]
```diff
@@ -0,0 +1,233 @@
...
+  string instance_id = 2;
+}
+
+// Defines a response for the attach cloud drive operation
+message SdkCloudDriveSetAttachResponse {
+  // AttachedDrives is the list of the drives that are attached
+  repeated AttachedCloudDrive attached_drives = 1;
+}
+
+// CloudDriveAttachedDrive represents a single attached drive
```
Comment (piyush-nimbalkar) [inline api/clouddrive.proto]: nit: *AttachedCloudDrive

File: api/clouddrive.proto [hunk 18]
```diff
@@ -0,0 +1,233 @@
...
+// Defines a response for the inspect cloud drive operation
+message SdkCloudDriveSetExpandResponse {
+  // CloudDriveSet that was requested to expand
+  CloudDriveSet cloud_drive_set = 1;
+}
+
+// Defines the operation types available to expand a drive set
+enum DriveSetExpandOperationType {
+  // Automatically pick the optimum operation type
+  Auto = 0;
```
Comment (piyush-nimbalkar) [inline api/clouddrive.proto]: nit: same here, all caps

File: api/clouddrive.proto [hunk 19]
```diff
@@ -0,0 +1,233 @@
...
+message SdkCloudDriveSetExpandRequest {
+  // ID of the cloud drive set to expand
+  string id = 1;
+  // PoolID of the pool inside the cloud drive to expand
+  string pool_id = 2;
+  // OperationType is the operation that's used to expand the drive set (optional)
+  DriveSetExpandOperationType operation_type = 3;
+}
+
+// Defines a response for the inspect cloud drive operation
```
Comment (piyush-nimbalkar) [inline api/clouddrive.proto]: *expand operation

File: api/api.proto [hunk 20]
```diff
@@ -232,6 +232,32 @@ message StoragePool {
...
+enum DriveCategory {
+  // Data indicates drive is being used for data purposes
+  data = 0;
```
[the remark this answers is not in the document]
Comment (piyush-nimbalkar) [reply api/api.proto]: Missed this one.

**Comment by ggriffiths** [comment]
Closing manually as we're closing old PRs before Jan 2020. Feel free to update and re-open if this is still relevant.

**Pull request closed without merging**